apps/studio/components/interfaces/Integrations/Queues/SingleQueue/QueueSettings.tsx

```diff
@@ -0,0 +1,303 @@
+                connectionString: project.connectionString,
+                revokes: [
+                  ...rolesNoLongerHavingPerms.map((x) => ({
+                    grantee: x,
+                    privilege_type: 'INSERT' as 'INSERT',
+                    relation_id: archiveTable.id,
+                  })),
+                  ...rolesNoLongerHavingPerms.map((x) => ({
+                    grantee: x,
+                    privilege_type: 'SELECT' as 'SELECT',
+                    relation_id: archiveTable.id,
+                  })),
+                ],
+              }),
+            ]
+          : []),
+        ...(grant.length > 0
+          ? [
+              grantPrivilege({
+                projectRef: project.ref,
+                connectionString: project.connectionString,
+                grants: grant.map((x) => ({
+                  grantee: x.role,
+                  privilege_type: x.action.toUpperCase(),
+                  relation_id: queueTable.id,
+                })) as TablePrivilegesGrant[],
+              }),
+              // Just grant select + insert on archive table as long as we're granting any perms to the queue table for the role
+              grantPrivilege({
+                projectRef: project.ref,
+                connectionString: project.connectionString,
+                grants: [
+                  ...rolesBeingGrantedPerms.map((x) => ({
+                    grantee: x,
+                    privilege_type: 'INSERT' as 'INSERT',
+                    relation_id: archiveTable.id,
+                  })),
+                  ...rolesBeingGrantedPerms.map((x) => ({
+                    grantee: x,
+                    privilege_type: 'SELECT' as 'SELECT',
+                    relation_id: archiveTable.id,
+                  })),
+                ],
+              }),
+            ]
+          : []),
+      ])
+      toast.success('Successfully updated permissions')
+      setOpen(false)
+    } catch (error: any) {
+      toast.error(`Failed to update permissions: ${error.message}`)
+    } finally {
+      setIsSaving(false)
+    }
+  }
+
+  useEffect(() => {
+    if (isSuccessPrivileges && queuePrivileges) {
+      const initialState = queuePrivileges.privileges.reduce((a, b) => {
+        return {
+          ...a,
+          [b.grantee]: { ...(a as any)[b.grantee], [b.privilege_type.toLowerCase()]: true },
+        }
+      }, {})
+      setPrivileges(initialState)
+    }
+  }, [isSuccessPrivileges])
+
+  return (
+    <Sheet open={open} onOpenChange={setOpen}>
+      <SheetTrigger asChild>
+        <ButtonTooltip
+          type="text"
+          className="px-1.5"
+          icon={<Settings />}
+          title="Settings"
+          onClick={() => setOpen(true)}
+          tooltip={{ content: { side: 'bottom', text: 'Queue settings' } }}
+        />
+      </SheetTrigger>
+      <SheetContent size="lg" className="overflow-auto flex flex-col gap-y-0">
+        <SheetHeader>
```

Comment:
The state of the sheet is kept if you close it via clicking outside.
Move all children of `SheetContent` in a separate component with all their state and fetching. The state will reset on each show.

Reply:
ahhhh good catch - hmm i'll opt to just reset the state in a useEffect for now